### ISSUE TITLE
Why is this better than using the specifier syntax?

### ISSUE THREAD

**Issue by domenic** [kmiller68/module-fallback-imports#2]
As you noted, there's a related proposal on the web side: https://github.com/drufball/layered-apis/#part-1-the-infrastructure

After some back and forth, we tentatively decided using the specifier syntax was the best approach there, as it automatically works in disparate contexts such as `src=""` and `import()` as well as `import` statements.

Although this version is a bit nicer aesthetically, are there any other advantages over using the specifier syntax? In the past when I've made similar TC39 proposals (e.g. for SRI), I was told to just use the specifier syntax, or out of band metadata, for this kind of thing.

**Comment by domenic** [comment]
You may also enjoy our exploration of different fallback syntaxes, which has some problem-space framing: https://docs.google.com/document/d/1jRQjQP8DmV7RL75u_67ps3SB1sjfa1bFZmbCMfJCvrM/edit

**Comment by kmiller68** [comment]
Unless the specifier syntax is adopted by TC-39 I don't see how there could ever be a way to implement a TC-39 native module. Thus all features related to TC-39 "owned" objects (Array, Set, Map, etc) are subject to backwards compatibility constraints in cumbersome ways. 

I would also argue that at some point the specifier syntax will become an unreadable mess if applied to too many things. For example, how do you parse:
src="std:math|https://myCDN.net/math.js?q=version=1.9.8{integrity=t5344ne3ohua4343rc498d9iu27id,myOtherAttributes=strict/magical/crazy}|https://myOtherCDN.net/math.js{integrity=t5344ne3ohua4343rc498d9iu27id,myOtherAttributes=words/404=gogogo}"

Obviously, that's pretty contrived but I think it's emblematic of the problem.

**Comment by domenic** [comment]
I don't think TC39 should be in the business of specifying module loading, but if they wanted to specify a built-in module, they could suggest a proposed specifier to hosts, and hosts would then integrate it into their own specifier pipelines. Hosts would then make specifier features such as fallbacks work on top of any type of module, built-in or not.

As for the aesthetic argument, I agree that `" else "` is slightly nicer than `|`, but I don't think that materially changes your example; it's just as unreadable a mess with the `" else "` version, IMO.

**Comment by kmiller68** [comment]
> I don't think TC39 should be in the business of specifying module loading, but if they wanted to specify a built-in module, they could suggest a proposed specifier to hosts, and hosts would then integrate it into their own specifier pipelines. Hosts would then make specifier features such as fallbacks work on top of any type of module, built-in or not.

I'm not proposing that TC-39 start specifying how module loading works. Only that our syntax support the idea there is more than one option. Even if this proposal makes it's way through TC-39 there would still need to be lifting done in W3C to actually use this feature. As for how TC-39 should provide native modules I totally agree with your system.

If you consider the following syntax (I'm not super sold that this has to be the syntax btw it's just for illustration):

```js
import "std:math", 
       "https://myCDN.net/math.js?q=version=1.9.8" with 
           {
                 integrity=t5344ne3ohua4343rc498d9iu27id, 
                 myOtherAttributes=strict/magical/crazy 
           },
       "https://myOtherCDN.net/math.js" with 
           {
                 integrity=t5344ne3ohua4343rc498d9iu27id,
                 myOtherAttributes=words/404=gogogo 
           };
```

I think this is very readable. With the specifier syntax you have no choice but to put everything on one line.

**Comment by domenic** [comment]
Ah, you were assuming the `with`. That got shot down pretty hard when I presented it to committee several meetings ago; see [my slide deck](https://docs.google.com/presentation/d/1qfoLTniLUVJ5YNFrha7BaVumAnW0ZgcCfUU8UbyyuYY/edit?usp=sharing) and contemporary notes. The committee's conclusion was instead we should use an out of band (presumably host-specific) mechanism for associating this information.

So I think it comes down to `" else "` vs. `|`. In both cases I think TC39 could specify built-in modules, and in both cases the host would provide a fallback. I'm still looking for the advantage. Is it purely aesthetic?

**Comment by kmiller68** [comment]
Interesting, I would have guessed that most websites would be using some system for slurping up their dependencies their tree wouldn't be that deep. I can see why out of band would make sense. I still think it might we worthwhile to add syntax but I'm beginning the think that html attributes might be the way to go.

**Comment by domenic** [comment]
Yeah, thinking further, syntax would be nice as long as we coordinated so there was something similar on the HTML and import() side. E.g.

```js
import "x" else "y", "z", "w";
import("x", { else: ["y", "z", "w"] });
```

```html
<script type=module src=x elsesrc="y z w"></script>
<script>
new Worker("x", { type: "module", else: ["y", "z", "w"] });
</script>
```

I'm not sure it's quite worth it compared to the simplicity of using the specifier in all four places, but it does look nicer. As long as everything was coordinated together I think we'd be pretty happy.

The alternative worth mentioning is moving even the fallbacks to an out of band config.
